src/debugger/Debugger.js

```diff
@@ -10,129 +10,155 @@
+      this.waitForRun(function(line) {
+        return line === "proceed";
       });
     }
   }
 
-  sendDebugInfo(ast: BabelNode, lineNum: number) {
-    this.channel.writeOut("breakpoint stopped " + lineNum);
-  }
-
-  parseCommand(command: string) {
+  executeCommand(command: string) {
     if (command.length === 0) {
       return;
     }
     let parts = command.split(" ");
     if (parts[0] === "breakpoint") {
-      this.parseBreakpointCommand(parts);
+      this.executeBreakpointCommand(this._parseBreakpointArguments(parts));
     }
   }
 
-  parseBreakpointCommand(parts: Array<string>) {
-    if (parts[1] === "add") {
-      let lineNum = parseInt(parts[2], 10);
-      let breakpoint = new Breakpoint(lineNum, true);
-      this.breakpoints[lineNum] = breakpoint;
-      this.channel.writeOut("added breakpoint " + lineNum);
-    } else if (parts[1] === "remove") {
-      let lineNum = parseInt(parts[2], 10);
-      invariant(lineNum in this.breakpoints);
-      delete this.breakpoints[lineNum];
-      this.channel.writeOut("removed breakpoint " + lineNum);
-    } else if (parts[1] === "enable") {
-      let lineNum = parseInt(parts[2], 10);
-      invariant(lineNum in this.breakpoints && !this.breakpoints[lineNum].enabled);
-      this.breakpoints[lineNum].enabled = true;
-      this.channel.writeOut("enabled breakpoint " + lineNum);
-    } else if (parts[1] === "disable") {
-      let lineNum = parseInt(parts[2], 10);
-      invariant(lineNum in this.breakpoints && this.breakpoints[lineNum].enabled);
-      this.breakpoints[lineNum].enabled = false;
-      this.channel.writeOut("disabled breakpoint " + lineNum);
+  executeBreakpointCommand(args: BreakpointCommandArguments) {
+    if (args.kind === "add") {
+      this.breakpoints.addBreakpoint(args.filePath, args.lineNum, args.columnNum);
+      this.channel.writeOut(`added breakpoint ${args.filePath} ${args.lineNum} ${args.columnNum}`);
+    } else if (args.kind === "remove") {
+      this.breakpoints.removeBreakpoint(args.filePath, args.lineNum, args.columnNum);
+      this.channel.writeOut(`removed breakpoint ${args.filePath} ${args.lineNum} ${args.columnNum}`);
+    } else if (args.kind === "enable") {
+      this.breakpoints.enableBreakpoint(args.filePath, args.lineNum, args.columnNum);
+      this.channel.writeOut(`enabled breakpoint ${args.filePath} ${args.lineNum} ${args.columnNum}`);
+    } else if (args.kind === "disable") {
+      this.breakpoints.disableBreakpoint(args.filePath, args.lineNum, args.columnNum);
+      this.channel.writeOut(`disabled breakpoint ${args.filePath} ${args.lineNum} ${args.columnNum}`);
     }
   }
 
+  _parseBreakpointArguments(parts: Array<string>): BreakpointCommandArguments {
+    invariant(parts[0] === "breakpoint");
+    let kind = parts[1];
+    let filePath = parts[2];
+    invariant(!isNaN(parts[3]));
+    let lineNum = parseInt(parts[3], 10);
+    let columnNum = 0;
+    if (parts.length === 5) {
+      invariant(!isNaN(parts[4]));
```

Comment:
parts[4] is a string. It makes no sense to check if it is NaN.

Reply:
Same as above